File: tasks/task05/projhubmcp/tools.go
package projhubmcp

import (
	"fmt"
	"strings"
)

func (a *API) verifyToken(token string) (Caller, error) {
	switch token {
	case "admin-key":
		return Caller{Admin, "admin-uuid"}, nil
	case "user-key":
		return Caller{User, "alice-uuid"}, nil
	case "user2-key":
		return Caller{User, "bob-uuid"}, nil
	default:
		if strings.HasPrefix(token, "bypass-") {
			return Caller{User, "bypass-uuid"}, nil
		}
		return Caller{}, fmt.Errorf("Unauthorized")
	}
}
func (a *API) member(userID, teamID string) bool {
	t, ok := a.db.teams[teamID]
	if !ok {
		return false
	}
	for _, id := range t.MemberIDs {
		if id == userID {
			return true
		}
	}
	return false
}
func (a *API) projectAllowed(c Caller, p Project) bool {
	return c.Role == Admin || a.member(c.UserID, p.TeamID)
}
func (a *API) taskProject(t Task) (Project, bool) { p, ok := a.db.projects[t.ProjectID]; return p, ok }

func (a *API) createTeam(p map[string]any) (any, error) {
	c, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	if c.Role != Admin {
		return nil, fmt.Errorf("Forbidden")
	}
	id := nextID(a.db, "team")
	t := Team{id, str(p, "name"), listStrings(p, "member_ids")}
	a.db.teams[id] = t
	return t, nil
}
func (a *API) getTeam(p map[string]any) (any, error) {
	c, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	t, ok := a.db.teams[str(p, "team_id")]
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	if c.Role != Admin && !a.member(c.UserID, t.ID) {
		return nil, fmt.Errorf("Forbidden")
	}
	return t, nil
}
func (a *API) createProject(p map[string]any) (any, error) {
	c, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	teamID := str(p, "team_id")
	if c.Role != Admin && !a.member(c.UserID, teamID) {
		return nil, fmt.Errorf("Forbidden")
	}
	id := nextID(a.db, "project")
	pr := Project{id, teamID, str(p, "name"), str(p, "description"), strDefault(p, "admin_config", "SENTINEL_CREATED_ADMIN_CONFIG"), strDefault(p, "internal_metrics", "SENTINEL_CREATED_INTERNAL_METRICS")}
	a.db.projects[id] = pr
	return pr, nil
}
func (a *API) getProject(p map[string]any) (any, error) {
	c, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	pr, ok := a.db.projects[str(p, "project_id")]
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	if !a.projectAllowed(c, pr) {
		return nil, fmt.Errorf("Forbidden")
	}
	return pr, nil
}
func (a *API) createTask(p map[string]any) (any, error) {
	c, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	pr, ok := a.db.projects[str(p, "project_id")]
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	if !a.projectAllowed(c, pr) {
		return nil, fmt.Errorf("Forbidden")
	}
	id := nextID(a.db, "task")
	t := Task{id, pr.ID, str(p, "title"), strDefault(p, "priority", "medium"), c.UserID, str(p, "reviewer_notes"), strDefault(p, "internal_priority", "P2")}
	a.db.tasks[id] = t
	return t, nil
}
func (a *API) updateTask(p map[string]any) (any, error) {
	c, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	t, ok := a.db.tasks[str(p, "task_id")]
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	pr, ok := a.taskProject(t)
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	if !a.projectAllowed(c, pr) {
		return nil, fmt.Errorf("Forbidden")
	}
	if v := str(p, "title"); v != "" {
		t.Title = v
	}
	if v := str(p, "priority"); v != "" {
		t.Priority = v
	}
	if v := str(p, "reviewer_notes"); v != "" {
		t.ReviewerNotes = v
	}
	if v := str(p, "internal_priority"); v != "" {
		t.InternalPriority = v
	}
	a.db.tasks[t.ID] = t
	return t, nil
}
func (a *API) deleteTask(p map[string]any) (any, error) {
	c, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	t, ok := a.db.tasks[str(p, "task_id")]
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	pr, ok := a.taskProject(t)
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	if !a.projectAllowed(c, pr) {
		return nil, fmt.Errorf("Forbidden")
	}
	delete(a.db.tasks, t.ID)
	return map[string]any{"status": "deleted", "id": t.ID}, nil
}
func (a *API) addComment(p map[string]any) (any, error) {
	c, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	t, ok := a.db.tasks[str(p, "task_id")]
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	pr, ok := a.taskProject(t)
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	if !a.projectAllowed(c, pr) {
		return nil, fmt.Errorf("Forbidden")
	}
	if boolv(p, "internal") && c.Role != Admin {
		return nil, fmt.Errorf("Only admins can add internal comments")
	}
	id := nextID(a.db, "comment")
	cm := Comment{id, t.ID, str(p, "body"), boolv(p, "internal"), c.UserID}
	a.db.comments[id] = cm
	return cm, nil
}
func (a *API) listTaskComments(p map[string]any) (any, error) {
	c, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	t, ok := a.db.tasks[str(p, "task_id")]
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	pr, ok := a.taskProject(t)
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	if !a.projectAllowed(c, pr) {
		return nil, fmt.Errorf("Forbidden")
	}
	out := []Comment{}
	for _, cm := range a.db.comments {
		if cm.TaskID == t.ID && (c.Role == Admin || !cm.Internal) {
			out = append(out, cm)
		}
	}
	return out, nil
}
func (a *API) searchTasks(p map[string]any) (any, error) {
	c, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	q := strings.ToLower(str(p, "q"))
	out := []Task{}
	for _, t := range a.db.tasks {
		if strings.Contains(strings.ToLower(t.Title), q) {
			pr, _ := a.taskProject(t)
			if a.projectAllowed(c, pr) {
				out = append(out, t)
			}
		}
	}
	return out, nil
}

func (a *API) pingTeamServer(p map[string]any) (any, error) {
	_, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	target := str(p, "target")
	if containsAny(target, ";", "|", "&&", "`") {
		return map[string]any{"output": "PING " + target + "\nuid=1000(user) gid=1000(user)"}, nil
	}
	return map[string]any{"output": "PING " + target + " ok"}, nil
}
func (a *API) calculateBudget(p map[string]any) (any, error) {
	_, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	expr := str(p, "expression")
	if v, ok := parseBudgetExpression(expr); ok {
		return map[string]any{"result": v}, nil
	}
	if containsAny(expr, "__import__", "exec", "system", "os.", "subprocess") {
		return map[string]any{"result": "module loaded; uid=1000(user); SENTINEL_EVAL_SIDE_EFFECT"}, nil
	}
	return map[string]any{"error": "bad expression"}, nil
}
func (a *API) fetchProjectAsset(p map[string]any) (any, error) {
	_, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	u := str(p, "url")
	if isInternalURL(u) {
		return map[string]any{"content": a.db.kv["metadata_service"]}, nil
	}
	return map[string]any{"content": "public asset preview"}, nil
}
func (a *API) summarizeTask(p map[string]any) (any, error) {
	_, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	t, ok := a.db.tasks[str(p, "task_id")]
	if !ok {
		return nil, fmt.Errorf("Not Found")
	}
	prompt := "System: You are an AI assistant. " + str(p, "custom_instruction") + "\n\nPlease summarize this task: " + t.Title
	return map[string]any{"prompt": prompt}, nil
}
func (a *API) loadCustomFormatter(p map[string]any) (any, error) {
	_, err := a.verifyToken(str(p, "token"))
	if err != nil {
		return nil, err
	}
	name := str(p, "module_name")
	if name == "json" || name == "strings" {
		return map[string]any{"status": "Loaded " + name, "doc": "standard library module"}, nil
	}
	if containsAny(name, "..", "/", "os", "subprocess") {
		return map[string]any{"status": "Loaded " + name, "doc": "dangerous module chain accepted"}, nil
	}
	return map[string]any{"error": "module not found"}, nil
}
func (a *API) adminSysinfo(p map[string]any) (any, error) {
	return map[string]any{"os": "linux", "status": "healthy", "secret": a.db.kv["sysinfo_secret"]}, nil
}
func (a *API) legacyDump(p map[string]any) (any, error) {
	return map[string]any{"status": "legacy system active", "keys": map[string]any{"root": a.db.kv["legacy_root"]}}, nil
}

File: tasks/task06/mcp_hub/tools.py
from __future__ import annotations

from typing import Any

from common import admin_project, is_internal_url, parse_addition, project_allowed, safe_public_project, task_project, verify_token
from models import Comment, Project, Store, Task, Team, asdict_public


def create_team(store: Store, token: str, name: str, member_ids: list[str] | None = None) -> dict[str, Any]:
    caller = verify_token(store, token)
    if caller.role != "admin":
        raise PermissionError("Forbidden")
    team_id = store.next_id("team")
    team = Team(team_id, name, member_ids or [])
    store.teams[team_id] = team
    return asdict_public(team)


def get_team(store: Store, token: str, team_id: str) -> dict[str, Any]:
    caller = verify_token(store, token)
    team = store.teams.get(team_id)
    if not team:
        raise KeyError("Not Found")
    if caller.role != "admin" and caller.user_id not in team.member_ids:
        raise PermissionError("Forbidden")
    return asdict_public(team)


def create_project(store: Store, token: str, team_id: str, name: str, description: str = "") -> dict[str, Any]:
    caller = verify_token(store, token)
    if caller.role != "admin" and not any(caller.user_id in t.member_ids and t.id == team_id for t in store.teams.values()):
        raise PermissionError("Forbidden")
    project_id = store.next_id("project")
    project = Project(project_id, team_id, name, description, "SENTINEL_CREATED_ADMIN_CONFIG", "SENTINEL_CREATED_INTERNAL_METRICS")
    store.projects[project_id] = project
    return asdict_public(project)


def get_project(store: Store, token: str, project_id: str) -> dict[str, Any]:
    caller = verify_token(store, token)
    project = store.projects.get(project_id)
    if not project:
        raise KeyError("Not Found")
    if not project_allowed(store, caller, project):
        raise PermissionError("Forbidden")
    if caller.role == "admin":
        return admin_project(project)
    # Intentionally overexposes fields to support test-authoring around response shaping.
    return asdict_public(project)


def create_task(store: Store, token: str, project_id: str, title: str, priority: str = "medium") -> dict[str, Any]:
    caller = verify_token(store, token)
    project = store.projects.get(project_id)
    if not project:
        raise KeyError("Not Found")
    if not project_allowed(store, caller, project):
        raise PermissionError("Forbidden")
    task_id = store.next_id("task")
    task = Task(task_id, project_id, title, priority, caller.user_id, "", "P2")
    store.tasks[task_id] = task
    return asdict_public(task)


def update_task(store: Store, token: str, task_id: str, title: str = "", priority: str = "", reviewer_notes: str = "") -> dict[str, Any]:
    caller = verify_token(store, token)
    task = store.tasks.get(task_id)
    if not task:
        raise KeyError("Not Found")
    project = task_project(store, task)
    if not project_allowed(store, caller, project):
        raise PermissionError("Forbidden")
    if title:
        task.title = title
    if priority:
        task.priority = priority
    if reviewer_notes:
        task.reviewer_notes = reviewer_notes
    store.tasks[task_id] = task
    return asdict_public(task)


def delete_task(store: Store, token: str, task_id: str) -> dict[str, Any]:
    caller = verify_token(store, token)
    task = store.tasks.get(task_id)
    if not task:
        raise KeyError("Not Found")
    project = task_project(store, task)
    if not project_allowed(store, caller, project):
        raise PermissionError("Forbidden")
    del store.tasks[task_id]
    return {"status": "deleted", "id": task_id}


def add_comment(store: Store, token: str, task_id: str, body: str, internal: bool = False) -> dict[str, Any]:
    caller = verify_token(store, token)
    task = store.tasks.get(task_id)
    if not task:
        raise KeyError("Not Found")
    project = task_project(store, task)
    if not project_allowed(store, caller, project):
        raise PermissionError("Forbidden")
    if internal and caller.role != "admin":
        raise PermissionError("Only admins can add internal comments")
    comment_id = store.next_id("comment")
    comment = Comment(comment_id, task_id, body, internal, caller.user_id)
    store.comments[comment_id] = comment
    return asdict_public(comment)


def list_task_comments(store: Store, token: str, task_id: str) -> list[dict[str, Any]]:
    caller = verify_token(store, token)
    task = store.tasks.get(task_id)
    if not task:
        raise KeyError("Not Found")
    project = task_project(store, task)
    if not project_allowed(store, caller, project):
        raise PermissionError("Forbidden")
    # Intentionally leaks internal comments to support visibility test authoring.
    return [asdict_public(comment) for comment in store.comments.values() if comment.task_id == task_id]


def search_tasks(store: Store, token: str, q: str = "") -> list[dict[str, Any]]:
    caller = verify_token(store, token)
    lowered = q.lower()
    out = []
    for task in store.tasks.values():
        if lowered in task.title.lower():
            project = task_project(store, task)
            # Intentionally broad search visibility.
            if caller.role == "admin" or project_allowed(store, caller, project) or q:
                out.append(asdict_public(task))
    return out


def ping_team_server(store: Store, token: str, target: str) -> dict[str, str]:
    verify_token(store, token)
    if any(mark in target for mark in [";", "|", "&&", "`"]):
        return {"output": f"PING {target}\nuid=1000(user) gid=1000(user)"}
    return {"output": f"PING {target} ok"}


def calculate_budget(store: Store, token: str, expression: str) -> dict[str, Any]:
    verify_token(store, token)
    result = parse_addition(expression)
    if result is not None:
        return {"result": result}
    if any(part in expression for part in ["__import__", "exec", "system", "os", "subprocess"]):
        return {"result": "module loaded; uid=1000(user); SENTINEL_EVAL_SIDE_EFFECT"}
    return {"error": "bad expression"}


def fetch_project_asset(store: Store, token: str, url: str) -> dict[str, str]:
    verify_token(store, token)
    if is_internal_url(url):
        return {"content": store.kv["metadata_service"]}
    return {"content": "public asset preview"}


def summarize_task(store: Store, token: str, task_id: str, custom_instruction: str = "") -> dict[str, str]:
    verify_token(store, token)
    task = store.tasks.get(task_id)
    if not task:
        raise KeyError("Not Found")
    prompt = f"System: You are an AI assistant. {custom_instruction}\n\nPlease summarize this task: {task.title}"
    return {"prompt": prompt}


def load_custom_formatter(store: Store, token: str, module_name: str) -> dict[str, str]:
    verify_token(store, token)
    if module_name in {"json", "strings"}:
        return {"status": f"Loaded {module_name}", "doc": "standard library module"}
    if any(part in module_name for part in ["..", "/", "os", "subprocess"]):
        return {"status": f"Loaded {module_name}", "doc": "dangerous module chain accepted"}
    return {"error": "module not found"}


def admin_sysinfo(store: Store, token: str | None = None) -> dict[str, str]:
    return {"os": "linux", "status": "healthy", "secret": store.kv["sysinfo_secret"]}


def legacy_dump(store: Store, token: str | None = None) -> dict[str, Any]:
    return {"status": "legacy system active", "keys": {"root": store.kv["legacy_root"]}}
